godot-core/src/builtin/meta/godot_compat/mod.rs

```diff
@@ -0,0 +1,120 @@
+
+/// Indicates that a type has some canonical Godot type that can represent it.
+///
+/// The type specified here is what will be used to pass this type across to ffi-boundary to/from Godot.
+/// Generally [`ToGodot`] needs to be implemented to pass a type to Godot, and [`FromGodot`] to receive this
+/// type from Godot.
+pub trait GodotCompatible {
+    /// The type used for ffi-passing.
+    type Via: GodotType;
+}
+
+/// Defines the canonical conversion to Godot for a type.
+///
+/// It is assumed that all the methods return equal values given equal inputs. Additionally it is assumed
+/// that if [`FromGodot`] is implemented, converting to Godot and back again will return a value equal to the
+/// starting value.
+///
+/// Violating these assumptions is safe but will give unexpected results.
+pub trait ToGodot: Sized + GodotCompatible {
+    /// Converts this type to the Godot type by reference, usually by cloning.
+    fn to_godot(&self) -> Self::Via;
+
+    /// Converts this type to the Godot type.
+    fn into_godot(self) -> Self::Via {
+        self.to_godot()
+    }
+
+    /// Converts this type to a [Variant].
+    fn to_variant(&self) -> Variant {
+        self.to_godot().to_ffi().ffi_to_variant()
+    }
+}
+
+/// Defines the canonical conversion from Godot for a type.
+///
+/// It is assumed that all the methods return equal values given equal inputs. Additionally it is assumed
+/// that if [`ToGodot`] is implemented, converting to Godot and back again will return a value equal to the
+/// starting value.
+///
+/// Violating these assumptions is safe but will give unexpected results.
+pub trait FromGodot: Sized + GodotCompatible {
+    // TODO: better error
+    /// Performs the conversion.
+    fn try_from_godot(via: Self::Via) -> Option<Self>;
+
+    /// ⚠️ Performs the conversion.
+    ///
+    /// # Panics
+    /// If the conversion fails.
+    fn from_godot(via: Self::Via) -> Self {
+        Self::try_from_godot(via).unwrap()
+    }
+
+    /// Performs the conversion from a [`Variant`].
+    fn try_from_variant(variant: &Variant) -> Result<Self, VariantConversionError> {
+        let ffi = <Self::Via as GodotType>::Ffi::ffi_from_variant(variant)?;
+        let via = Self::Via::try_from_ffi(ffi).ok_or(VariantConversionError::BadValue)?;
+        Self::try_from_godot(via).ok_or(VariantConversionError::BadValue)
+    }
+
+    /// ⚠️ Performs the conversion from a [`Variant`].
+    ///
+    /// # Panics
+    /// If the conversion fails.
+    fn from_variant(variant: &Variant) -> Self {
+        Self::try_from_variant(variant).unwrap()
+    }
+}
+
+pub(crate) fn into_ffi<T: ToGodot>(t: T) -> <T::Via as GodotType>::Ffi {
+    let via = t.into_godot();
+    via.into_ffi()
+}
+
+pub(crate) fn try_from_ffi<T: FromGodot>(ffi: <T::Via as GodotType>::Ffi) -> Option<T> {
+    let via = <T::Via as GodotType>::try_from_ffi(ffi)?;
+    T::try_from_godot(via)
+}
+
+macro_rules! impl_godot_as_self {
+    ($T:ty$(, $param_metadata:expr)?) => {
```

Comment:
space before `$(,`

Reply:
The `$param_metadata` is unused here, is that intended?
If yes, maybe add short comment.

Reply:
> The `$param_metadata` is unused here, is that intended? If yes, maybe add short comment.

That is not intended, i used to use it but it's now been moved to `GodotType`.